src/main/java/org/myrobotlab/service/ProgramAB.java

```diff
@@ -798,28 +678,69 @@ public void writeAIML() {
    */
   public void writeAndQuit() {
     // write out all bots aiml & save all predicates for all sessions?
-    for (Bot bot : bots.values()) {
-      if (bot == null) {
-        log.info("no bot - don't need to write and quit");
-        continue;
-      }
-      try {
-        savePredicates();
-        // important to save learnf.aiml
-        writeAIML();
-      } catch (IOException e1) {
-        log.error("saving predicates threw", e1);
+    for (BotInfo bot : bots.values()) {
+      if (bot.isActive()) {
+        try {
+          savePredicates();
+          // important to save learnf.aiml
+          writeAIML();
+          bot.writeQuit();
+        } catch (IOException e1) {
+          log.error("saving predicates threw", e1);
+        }
       }
-      bot.writeQuit();
     }
   }
 
-  public String setPath(String path) {
-    if (path != null && !path.equals(this.path)) {
-      this.path = path;
+  /**
+   * Verifies and adds a new path to the search directories for bots
+   * 
+   * @param path
+   * @return
+   */
+  public String addBotPath(String path) {
+    // verify the path is valid
+    File botPath = new File(path);
+    File verifyAiml = new File(FileIO.gluePaths(path, "aiml"));

```

Comment:
I thought we were going to drop the requirement for having an aiml subdirectory...

Reply:
I'd like one indicator besides failing when lazy-loading creating the bot, to give a user some warning that it won't work.  Since the new Bot(path) in programab-kw has hardcoded sub-dirs, a subdir of aiml seemed to be an "ok" test

Reply:
This isn't a warning..  this will prevent you from creating a bot that doesn't have any aiml yet.  I think we should remove this verification.

Reply:
So empty dir as a start?, yet Bots actually do require convention - they have hardcoded sub-directories. 

If its only a empty directory and i can dump an aiml file named anything and get some kind of cause and effect, I'd consider taking the validation out.